### DISCUSSION TITLE
How to validate api key on gateway?

### DISCUSSION THREAD

**Discussion by BCantos17** [apollographql/apollo-server#6774] [Q&A] [answered]
Have an Apollo Gateway app and cannot find a documentation on how to inspect the header in order to validate an API Auth Key and then based on that validation, allow the query to continue. There is a `introspectionHeaders` method but that is deprecated and I have a federated graph so introspection in general seems out of the question.

Seems like `willSendRequest` is suppose to do something like that but all I found when googling was how to modify the context to add to the header and not to validate a pre-existing key in the header to allow to query to continue.

Is something like this possible?

**Comment by glasser** [comment, marked as answer]
One option is just to do this as a middleware in your Express (or whatever) app before the Apollo Server middleware.

In theory, your context function might be a good place to do this, or requestWillStart, but I'll admit that error handling around these functions is not ideal in Apollo Server (especially in AS3; maybe a bit better in the forthcoming AS4).

**Reply by BCantos17** [reply]
Yeah I thought of that but we would have to implement that check on all of our services which can and will get tedious as we continue to add more and more subgraphs. Wanted to check if there was a way to do this check on the gateway level

**Reply by glasser** [reply]
Sure, I meant in the gateway.

**Reply by BCantos17** [reply]
Oh, my bad! I am not a nodejs developer and not too familiar with some terminology. I am using the default ApolloServer class which looks like it's using express underneath. After some quick googling I came across [this](https://www.geeksforgeeks.org/basic-authentication-in-node-js-using-http-header/). Is this what you were proposing?

**Reply by glasser** [reply]
Take a look at https://www.apollographql.com/docs/apollo-server/integrations/middleware/#swapping-out-apollo-server

**Reply by BCantos17** [reply]
Thank you, this is helpful! I am going to attempt this and will get back on the results

**Reply by BCantos17** [reply]
This worked but not _exactly_ as I expected it to. I cannot reach the landing page for the gateway without the api key. I think this will do for my current needs but I was wondering if there are/were any discussions about doing something like this on the gateway level so that way I can go to the landing page but not make any queries without adding the api key in the header?

**Reply by glasser** [reply]
Might not be want you want to hear, but: you can mimic the [logic that Apollo Server does to decide if it's serving the landing page](https://github.com/apollographql/apollo-server/blob/d171c0587603ff910d9763a53b062675f53a6881/packages/apollo-server-express/src/ApolloServer.ts#L215-L224).  (Basically based on the `accept` header)

**Reply by BCantos17** [reply]
that works! It's not my ideal solution but this does exactly what I expected. Thanks a lot
